src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/LSPRazorProjectHost.cs

```diff
@@ -0,0 +1,124 @@
+        private readonly LSPEditorFeatureDetector _featureDetector;
+        private readonly List<Lazy<ILanguageClient, ILanguageClientMetadata>> _applicableLanguageClients;
+        private readonly Lazy<ILanguageClientBroker> _languageClientBroker;
+
+        [ImportingConstructor]
+        public LSPRazorProjectHost(
+            JoinableTaskContext joinableTaskContext,
+            LSPEditorFeatureDetector featureDetector,
+            Lazy<ILanguageClientBroker> languageClientBroker,
+            [ImportMany] IEnumerable<Lazy<ILanguageClient, IDictionary<string, object>>> languageClients)
+        {
+            if (joinableTaskContext is null)
+            {
+                throw new ArgumentNullException(nameof(joinableTaskContext));
+            }
+
+            if (featureDetector is null)
+            {
+                throw new ArgumentNullException(nameof(featureDetector));
+            }
+
+            if (languageClientBroker is null)
+            {
+                throw new ArgumentNullException(nameof(languageClientBroker));
+            }
+
+            if (languageClients is null)
+            {
+                throw new ArgumentNullException(nameof(languageClients));
+            }
+
+            _joinableTaskFactory = joinableTaskContext.Factory;
+            _featureDetector = featureDetector;
+            _languageClientBroker = languageClientBroker;
+            _applicableLanguageClients = GetApplicableClients(languageClients);
+        }
+
+        public async Task LoadAsync()
+        {
+            await _joinableTaskFactory.SwitchToMainThreadAsync();
+
+            if (!_featureDetector.IsLSPEditorFeatureEnabled())
+            {
+                return;
+            }
+
+            foreach (var client in _applicableLanguageClients)
+            {
+                _languageClientBroker.Value.LoadAsync(client.Metadata, client.Value).Forget();
+            }
+        }
+
+        public Task UnloadAsync()
+        {
+            return Task.CompletedTask;
+        }
+
+        private static List<Lazy<ILanguageClient, ILanguageClientMetadata>> GetApplicableClients(IEnumerable<Lazy<ILanguageClient, IDictionary<string, object>>> languageClients)
+        {
+            var applicableContentTypes = new[]
+            {
+                RazorLSPConstants.RazorLSPContentTypeName,
+                RazorLSPConstants.CSharpLSPContentTypeName,
+                RazorLSPConstants.HtmlLSPContentTypeName,
+            };
+
+            var applicableClients = new List<Lazy<ILanguageClient, ILanguageClientMetadata>>();
+            foreach (var client in languageClients)
+            {
+                if (!client.Metadata.TryGetValue(nameof(ILanguageClientMetadata.ContentTypes), out var contentTypeValue) ||
+                    !(contentTypeValue is IEnumerable<string> contentTypes) ||
+                    !contentTypes.Intersect(applicableContentTypes).Any())
```

Comment:
>contentTypes.Intersect(applicableContentTypes).Any() [](start = 21, length = 52)

.Contains?

Reply:
I don't believe that'll be possible -> `Contains` takes a `string` ContentType, while `applicableContentTypes` is a `string[]`.

Reply:
Gotcha, I missed that applicableContentTypes was an array

---
In reply to: [450419680](https://github.com/dotnet/aspnetcore-tooling/pull/2164#discussion_r450419680) [](ancestors = 450419680)